tech_reports/LLMs/llms.md

```diff
@@ -55,14 +55,294 @@ Authors:
+- A tensor of size `(seqlen, dim)` in DRAM for prefill.
+
+Each attention test compares the attention module output and KV-cache correlation between the PyTorch host implementation and the TTNN device implementation.
+
+The current version of the MLP module class (`TtLlamaMLP`) handles prefill and decode in the same file but has some technical differences (mentioned in the section below).
+
+The decoder module (which encapsulates both attention and MLP) and model module (which encapsulates the decoder and the remaining parts of the Llama3 model) also handle prefill and decode in the same file, but they call the respective modes within the attention and MLP modules.
+
+#### **Technical Implementation Differences**
+
+The intermediate activations in prefill mode are kept in DRAM, due to the large size of the tensors which contain the entire sequence length. In decode mode, the intermediate activations are kept in L1 memory instead, since in this mode the sequence length to compute is just 1 (one token at the time), reducing latency.
+
+##### 1. Reshaping for Large Matrix Multiplications
+
+Please see the [attention source code](../../models/demos/llama3/tt/llama_attention.py) for reference.
+
+In prefill mode, when the input sequence length is very large, the model reshapes its input tensors to process sequences in smaller chunks in parallel for larger matrix multiplications, such as `wqkv`, `wo` in the attention module, and `w1`, `w2`, `w3` in the MLP module. This reshaping prevents running out of memory in cases of long prefill sequence lengths. For instance:
+
+```python
+if seq_len > 2048:
+    x_11SH = ttnn.reshape(x_11SH, [1, seq_len // 2048, 2048, -1])
+
+xqkv_fused = ttnn.linear(
+    x_11SH,
+    self.wqkv,
+    dtype=ttnn.bfloat16,
+    memory_config=ttnn.DRAM_MEMORY_CONFIG,
+    compute_kernel_config=self.compute_kernel_config_hifi2,
+    program_config=self.model_config["XQKV_PREFILL_PROGCFG"](seq_len),
+)
+```
+
+This reshaping is not needed for decode mode because it only processes one token at a time. Instead, the parallelization for decode mode is done over user batches, which currently only goes up to 32.
+
+##### 2. KV Cache Management
+
+The KV-cache is filled during prefill using the `ttnn.experimental.paged_fill_cache` operation. This supports page tables, which enables the hot-swapping of new users when the full model is deployed.
+
+```python
+# Fill cache with initial states
+ttnn.experimental.paged_fill_cache(
+    keys_BKSD,
+    k_fill,
+    page_table,
+    batch_idx=user_id
+)
+```
+
+Similarly, during decode, the KV-cache update is done by `ttnn.experimental.paged_update_cache`, which updates the new KV values for all the users currently processing, with their respective positions.
+
+```python
+# Update KV cache with a single new token
+ttnn.experimental.paged_update_cache(
+    keys,
+    k_heads_1BKD,
+    update_idxs_tensor=current_pos,
+    page_table=page_table
+)
+```
+
+##### 3. Attention Computation
```

Comment:
This section feels a bit weird that we don't have a sentence before showing the code, but I couldn't come up with anything that would fit nicely, so I think it's probably ok.